collector/fixtures/e2e-output.txt

```diff
@@ -785,43 +788,82 @@ node_md_blocks_synced{device="md10"} 3.14159265e+08
 node_md_blocks_synced{device="md11"} 4.190208e+06
 node_md_blocks_synced{device="md12"} 3.886394368e+09
 node_md_blocks_synced{device="md127"} 3.12319552e+08
+node_md_blocks_synced{device="md13"} 4.88383936e+08
+node_md_blocks_synced{device="md20"} 9.76751616e+08
 node_md_blocks_synced{device="md219"} 7932
 node_md_blocks_synced{device="md3"} 5.853468288e+09
 node_md_blocks_synced{device="md4"} 4.883648e+06
+node_md_blocks_synced{device="md50"} 8.790400512e+09
 node_md_blocks_synced{device="md6"} 1.6775552e+07
 node_md_blocks_synced{device="md7"} 7.813735424e+09
 node_md_blocks_synced{device="md8"} 1.6775552e+07
 node_md_blocks_synced{device="md9"} 523968
-# HELP node_md_disks Total number of disks of device.
+# HELP node_md_disks Number of disks of device. (state="any" equals to the number of disks in the array)
```

Comment:
I think to parse the contents of /proc/mdstat is not a simple task (not for a human, let alone putting every possible notation in regexes).

Doing this, we may very well cross the line that is drawn in the last section of https://github.com/prometheus/node_exporter/blob/0126aca1382d1700066f45692eb0a54b4523a052/CONTRIBUTING.md 
```
The metrics should not get transformed in a way that is hardware specific and 
would require maintaining any form of vendor based mappings or conditions. 
```

But maybe I misread that, so let's ignore that for a moment and look at the examples we have on hand.

I've commented the /proc/mdstat file inline below and want to make sure we are on the same page here.

The most important cases for any metric regarding mdstat are the ones where something goes upside down. Unfortunately that are the cases where mdstat starts to lag information (probably just because it does not know certain things at that time). And that's the time when mdstat looks different from the normal cases, making it harder to build a parser.

Therefore, reading `/sys/block/md*` is most likely the way we should go.
(Note: the information that I miss in `/proc/mdstat` is probably neither in 
`/sys/block/md*`) But for that interface at least we do not need regex matching.

Example:
```php
/sys/block/md0/md/array_state clean
/sys/block/md0/md/degraded 1
/sys/block/md0/md/level raid1
/sys/block/md0/md/raid_disks 4
```

Now to the special cases where vital information is missing and has to be made up:

1. md6:
There is 1 out of 2 disks active. The other disk gets recovered.
This rebuilding state can be easily derived from the very existence of the 
recovery line and therefore a state `state=rebuilding` or something alike, 
with a value of 1 is easy to put in (`node_md_disks{device="md6",state="rebuilding"} 1`).
2. md7:
There are 3 out of 4 disks active. In difference to case md6 there is no rebuild in progress.
Therefore we can assume by calculating the difference that there is one disk missing (`node_md_disks{device="md6",state="missing"} 1 `).
3. md2:
*Now it gets hairy*. No knowledge about: raid level, number of disks, failed or missing disks.
I do not have such a degraded raid on hand, so I would have to make one to see what is there to get out of `/sys/block/md2`. But I hardly doubt that `/proc/mdstat` gives sufficient information here and I suspect even in the sys filesystem will be gaps.
**Any hints how to handle this case are welcome!**
4. md11:
This has an example why parsing text is bad and we should read /sys/block/md*
5. md219:
*No way to tell how many disks there should be active.*
6. md13:
The missing drive can be calculated.
7. md20:
This is the example where the whole is greater than the sum of its parts. ;-).
8. Why do we have those odd md naming in our examples?

**Roundup:**
It might be better to leave the failed, missing and rebuilding states out and advise users to alert on decrease of active disks or something like that. If we can not ensure that we report missing or failing disks in every case, we will most likely let people down who are going to alert on increasing numbers of failed/missing.

**commented /proc/mdstat**
```viml
Personalities : [linear] [multipath] [raid0] [raid1] [raid6] [raid5] [raid4] [raid10]
md3 : active raid6 sda1[8] sdh1[7] sdg1[6] sdf1[5] sde1[11] sdd1[3] sdc1[10] sdb1[9]
      5853468288 blocks super 1.2 level 6, 64k chunk, algorithm 2 [8/8] [UUUUUUUU]
node_md_disks{device="md3",state="active"} 6
node_md_disks{device="md3",state="any"} 6
node_md_disks{device="md3",state="failed"} 0
node_md_disks{device="md3",state="spare"} 0


md127 : active raid1 sdi2[0] sdj2[1]
      312319552 blocks [2/2] [UU]
node_md_disks{device="md123",state="active"} 2
node_md_disks{device="md123",state="any"} 2
node_md_disks{device="md123",state="failed"} 0
node_md_disks{device="md123",state="spare"} 0
*TODO*: I think this case is a duplicate of md0


md0 : active raid1 sdi1[0] sdj1[1]
      248896 blocks [2/2] [UU]
node_md_disks{device="md0",state="active"} 2
node_md_disks{device="md0",state="any"} 2
node_md_disks{device="md0",state="failed"} 0
node_md_disks{device="md0",state="spare"} 0


md4 : inactive raid1 sda3[0] sdb3[1]
      4883648 blocks [2/2] [UU]
node_md_disks{device="md4",state="active"} 2
node_md_disks{device="md4",state="any"} 2
node_md_disks{device="md4",state="failed"} 0
node_md_disks{device="md4",state="spare"} 0


md6 : active raid1 sdb2[2] sda2[0]
      195310144 blocks [2/1] [U_]
      [=>...................]  recovery =  8.5% (16775552/195310144) finish=17.0min speed=259783K/sec
node_md_disks{device="md6",state="active"} 1
node_md_disks{device="md6",state="any"} 2
node_md_disks{device="md6",state="failed"} 0
node_md_disks{device="md6",state="spare"} 0
node_md_disks{device="md6",state="rebuilding"} 1 *TODO*


md8 : active raid1 sdb1[1] sda1[0]
	  195310144 blocks [2/2] [UU]
	  [=>...................]  resync =  8.5% (16775552/195310144) finish=17.0min speed=259783K/sec
node_md_disks{device="md8",state="active"} 2
node_md_disks{device="md8",state="any"} 2
node_md_disks{device="md8",state="failed"} 0
node_md_disks{device="md8",state="spare"} 0


md7 : active raid6 sdb1[0] sde1[3] sdd1[2] sdc1[1]
      7813735424 blocks super 1.2 level 6, 512k chunk, algorithm 2 [4/3] [U_UU]
      bitmap: 0/30 pages [0KB], 65536KB chunk
node_md_disks{device="md7",state="active"} 3
node_md_disks{device="md7",state="any"} 4
node_md_disks{device="md7",state="failed"} 0
node_md_disks{device="md7",state="spare"} 0
node_md_disks{device="md6",state="missing"} 1 *TODO*


md9 : active raid1 sdc2[2] sdd2[3] sdb2[1] sda2[0]
      523968 blocks super 1.2 [4/4] [UUUU]
	resync=DELAYED
node_md_disks{device="md9",state="active"} 4
node_md_disks{device="md9",state="any"} 4
node_md_disks{device="md9",state="failed"} 0
node_md_disks{device="md9",state="spare"} 0


md10 : active raid0 sda1[0] sdb1[1]
      314159265 blocks 64k chunks
node_md_disks{device="md10",state="active"} 2
node_md_disks{device="md10",state="any"} 2
node_md_disks{device="md10",state="failed"} 0
node_md_disks{device="md10",state="spare"} 0


*BIG TODO*
md2 : inactive sda6[1](S) *TODO* - I know this is from a failed raid0 but where to get that from?
      1938311476 blocks super 1.2
node_md_disks{device="md2",state="active"} 0
node_md_disks{device="md2",state="any"} 2 *TODO* should be 2 - but thats only implicit knowledge as raid0 is always 2? *WARNING*: see case md00
node_md_disks{device="md2",state="failed"} 1 *TODO* - FAILED RAID0 - where to get that from?
node_md_disks{device="md2",state="spare"} 1
node_md_disks{device="md2",state="missing"} 1 *TODO* - FAILED RAID0 - shall it be failed or missing?


md11 : active (auto-read-only) raid1 sdb2[0] sdc2[1]
      4190208 blocks super 1.2 [2/2] [UU]
	resync=PENDING
node_md_disks{device="md11",state="active"} 2
node_md_disks{device="md11",state="any"} 2
node_md_disks{device="md11",state="failed"} 0
node_md_disks{device="md11",state="spare"} 0
*TODO*: this case has an "(auto-read-only)" flag set - can a raid0 go into "(auto-read-only)"? if so, we have to account for that in the way we count drives.


md12 : active raid0 sdc2[0] sdd2[1]
      3886394368 blocks super 1.2 512k chunks
node_md_disks{device="md12",state="active"} 2
node_md_disks{device="md12",state="any"} 2
node_md_disks{device="md12",state="failed"} 0
node_md_disks{device="md12",state="spare"} 0


md219 : inactive sdb[2](S) sdc[1](S) sda[0](S)
    7932 blocks super external:imsm
node_md_disks{device="md219",state="active"} 0
node_md_disks{device="md219",state="any"} 3 *TODO* - where to get a this from? we don't know if there are drives missing.
node_md_disks{device="md219",state="failed"} 0
node_md_disks{device="md219",state="spare"} 3


md00 : active raid0 xvdb[0]
      4186624 blocks super 1.2 256k chunks
node_md_disks{device="md00",state="active"} 1
node_md_disks{device="md00",state="any"} 1
node_md_disks{device="md00",state="failed"} 0
node_md_disks{device="md00",state="spare"} 0
*NOTE*: this seems to be the JBOD mode with just one drive, if someone wonders like me at first


md50 : active raid5 sda1[5] sdc1[0] sdb1[4](F) sdd1[1]
      8790400512 blocks super 1.2 level 5, 512k chunk, algorithm 2 [4/3] [UUU_]
node_md_disks{device="md50",state="active"} 3
node_md_disks{device="md50",state="any"} 4
node_md_disks{device="md50",state="failed"} 1
node_md_disks{device="md50",state="spare"} 0
node_md_disks{device="md50",state="missing"} 1 *TODO*


md13 : active raid1 sde1[6](F) sdg1[1] sdb1[4] sdd1[3] sdc1[2]
      488383936 blocks [6/4] [_UUUU_]
node_md_disks{device="md13",state="active"} 4
node_md_disks{device="md13",state="any"} 6
node_md_disks{device="md13",state="failed"} 1
node_md_disks{device="md13",state="spare"} 0
node_md_disks{device="md50",state="missing"} 1 TODO


md20 : active raid5 sdh1[5](S) sdg1[4] sdf1[3] sde1[2] sdd1[1] sdc1[0]
      976751616 blocks level 5, 64k chunk, algorithm 2 [5/5] [UUUUU]
      bitmap: 3/233 pages [12KB], 512KB chunk
node_md_disks{device="md20",state="active"} 5
node_md_disks{device="md20",state="any"} 5
node_md_disks{device="md20",state="failed"} 0
node_md_disks{device="md20",state="spare"} 1
*TODO*: this is a case where it does not sum up because of the spare disk.

unused devices: <none>
```